Fix uprobe listing and attaching when target is in a separate mount namespace

dalehamel opened the pull request:
Fixes https://github.com/iovisor/bpftrace/issues/968, at least for uprobes.

I'll follow up with a separate issue and patches for USDT probes, which I think are being similarly mishandled when the target is in a separate mount namespace.

I wrote two new regression tests, which utilize a wrapper  (mountns_wrapper.c) that I made for the runtime tests which will create a new mount namespace, and privately bind mount the test program folder there. This simulates a docker container somewhat, as the path won't be visible to bpftrace, which is in a separate mount namespace. This is, I think, about the smallest, simplest wrapper I could use to reproduce the issue.

Both regression tests fail without my changes, and pass with them.

My changes are pretty simple:

* In the semantic analyzer, if a PID is specified, then the target path has /proc/PID/root prepended to it. This allows attaching to probes in where the binary being uprobed is in a different mount namespace. There don't seem to be any side effects to just prepending the /proc relative root path, as all of the other tests including runtime tests continue to pass.
* In the list code, there is already special handling for if a PID is specified. I just make the absolute path include the /proc/PID/root bits at the front, and it allows listing probes for a binary in a separate mount namespace.

@danobi @fbs @mmarchini please take a look when you have a chance

fbs commented on src/ast/semantic_analyser.cpp:
```diff
@@ -1240,7 +1240,14 @@ void SemanticAnalyser::visit(AttachPoint &ap)
+
+    if (bpftrace_.pid_ > 0) {
+      std::ostringstream abs;
+      abs << "/proc/" << bpftrace_.pid_ << "/root/" << paths.front();
```
should we make a function that does this? Makes it easier to reuse/refactor when needed.

dalehamel replied:
Thanks for the suggestion, I imagine this will probably be needed elsewhere at some point.

I put it in util.cpp

fbs commented on tests/testprogs/mountns_wrapper.c:
```diff
@@ -0,0 +1,48 @@
+    if (mount(NULL, "/", NULL, MS_REC | MS_PRIVATE, NULL) != 0)
+      errExit("Failed to make mount private");
+
+    if (mkdir(private_mount, S_IRWXG) != 0 && (errno != EEXIST))
```
nit: personally prefer octal file modes (0770) over these weird S_IRWXG enums

dalehamel replied:
changed, better to be consistent. Thanks!

fbs commented on tests/testprogs/mountns_wrapper.c:
```diff
@@ -0,0 +1,48 @@
+      errExit("Failed to make private mount dir");
+
+    char path[BUFFER_SIZE];
+    int size = readlink("/proc/self/exe", path, sizeof(path));
```
`BUFFER_SIZE - 1` as your're writing to `path[size]` below.

dalehamel replied:
thanks

fbs commented on tests/testprogs/mountns_wrapper.c:
```diff
@@ -0,0 +1,48 @@
+int main(int argc, char *argv[])
+{
+
+    char* private_mount = "/tmp/unshare/";
```
I don't think this is unique enough to be run on peoples machines. `/tmp/bpftrace-unshare-test` is more likely to prevent conflicts and people will know where the file came from.

dalehamel replied:
I'll make sure it is unique too.

I don't think any of the other tests use /tmp as far as I know, if they do maybe a shared subfolder per test run could be a future standard?

dalehamel commented:
> Some quick comments

Thanks @fbs , sorry the C for mountns_wrapper was a little sloppy. Cleaned it up and ran through clang-format too, added some docs.

> Havent looked at the details of the unshare/mount calls yeet.

I added some comments, the tl;dr is that the unshare creates a new mount namespace, the namespace sets itself to private mount propagation, and then the testprog folder is mounted in a temporary directory, where it is then called from with execvp, replacing the mountns_wrapper process with the usual test program.

fbs commented:
Minor but:

```
$ sudo ./mountns_wrapper ./uprobe_test &
$ sudo bpftrace -l -p $(pidof uprobe_test)  'u:*'

uprobe:/proc/13684/root//uprobe_test:__new_fopen
uprobe:/proc/13684/root//uprobe_test:__wmemcpy
```

The double `//` in here is a bit ugly.

Other than that it looks fine, although I'm no namespace expert.

> I'll follow up with a separate issue and patches for USDT probes, which I think are being similarly mishandled when the target is in a separate mount namespace.

Yse looks like it (on a static binary):

```
$ sudo bpftrace -l -p $(pidof uprobe_test)  | grep usdt
usdt:/uprobe_test:libc:memory_mallopt_arena_max
usdt:/uprobe_test:libc:memory_mallopt_arena_test
```

dalehamel commented:
> The double // in here is a bit ugly.

I agree, I'll fix it.

dalehamel commented:
It looks like the behavior of this code changed quite a lot, I'm working on a rebase

dalehamel commented:
@fbs I addressed your comments on the double path separator

While I was working on this @acj landed 5005902 and 3f7a94a which created conflicts. @acj you might want to look this over / comment.

Because of 3f7a94a, I reworked this to add a new interface to `resolve_binary_path`, overloading it to allow optionally specifying a PID. If the PID is specified AND is in a different mount namespace, then
PATH is read based on the PID's own environment, rather than bpftrace's.

I couldn't figure out a good way to add a unit test for the new interface, as it requires spawning a child process and knowing its pid. I thought of doing this with `exec_system`, but as it stands the runtime tests cover this behavior.

I had to write a new helper function as well, to determine if the target id is in a different mount namespace.

The resulting refactor I think is actually more correct and a better iteration than the original, as it now only appends the /proc/PID/root bit if it knows for certain the target pid needs this.

dalehamel commented on src/list.cpp:
```diff
@@ -152,7 +152,7 @@ void list_probes(const BPFtrace &bpftrace, const std::string &search_input)
     {
       executable = get_pid_exe(bpftrace.pid_);
-      absolute_exe = executable;
+      absolute_exe = path_for_pid_mountns(bpftrace.pid_, executable);
```
this should maybe be changed to resolve_binary_path, wdyt?

dalehamel replied:
i think i'll leave it as is for now, as this should always be safe to take the path relative to the pid's mount namespace, whether the mount namespace is the same as for bpftrace or note.

dalehamel replied:
yeah nvm, resolve_binary_path returns a vector.

dalehamel commented on src/utils.cpp:
```diff
@@ -521,13 +566,69 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
+*/
+static bool pid_in_different_mountns(int pid) {
+
+  const size_t BUFSIZE = 4096;
```
I agonized over this one. Realistically I don't expect this to ever exceed ever 32 bytes, as it is based on a formatted print of a known size, bounded by the the maximum PID size, which is currently 9 digits (4 million is max pid).

Likewise, reading the symlink should always return the strrep of an inode ID for the mount ns, again less than 32 bytes.

However, theoretically as these pare paths, i went with PATH_MAX. It's huge overkill, but probably still so fast to stack allocate that it doesn't matter.

dalehamel replied:
thanks for weighing in, i'll stick with 64.

dalehamel commented on src/utils.cpp:
```diff
@@ -77,6 +77,10 @@ static bool provider_cache_loaded = false;
 // Maps all providers of pid to vector of tracepoints on that provider
 static std::map<std::string, usdt_probe_list> usdt_provider_cache;
 
+static bool pid_in_different_mountns(int pid);
```
I made this static for now because it hasn't been needed elsewhere, but this might be useful in other places in the code.

dalehamel commented on src/utils.cpp:
```diff
@@ -508,10 +511,53 @@ std::string exec_system(const char* cmd)
+    pid_environ_path << "/proc/" << pid << "/environ";
+    std::ifstream environ(pid_environ_path.str());
+
+    if(environ) {
```
I ran the patch through clang-format, per #639 to try and avoid style nits like this

acj replied:
Sorry, I deleted this comment after seeing your second commit, but GH kept it for some reason. Please disregard.

dalehamel replied:
All good, I'm looking forward to #639 being merged ;)

dalehamel commented on src/utils.cpp:
```diff
@@ -521,13 +566,69 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
+    if (bytes > 0 && bytes < BUFSIZE)
+      pid_ns[bytes] = '\0';
+    else
+      return false;
```
Done, I rolled this up with the check for errno as well.

dalehamel commented on src/utils.cpp:
```diff
@@ -521,13 +566,69 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
+  else
+    return false;
+
+  sprintf(pid_path, "/proc/%d/ns/mnt", pid);
```
done

dalehamel commented on src/utils.cpp:
```diff
@@ -521,13 +566,69 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
+    return false;
+
+  sprintf(pid_path, "/proc/%d/ns/mnt", pid);
+  if (access(pid_path, F_OK) != -1) {
```
good catch, I've dropped the `access` check and will just check errno when I check the edge cases on `bytes`

dalehamel commented on tests/testprogs/mountns_wrapper.c:
```diff
@@ -0,0 +1,78 @@
+    errExit("Must specify test program as only argument.");
+
+  // Enter a new mount namespace
+  if (unshare(CLONE_NEWNS | CLONE_FS) != 0)
```
good catch, saw than in `man 2 unshare` when I just checked:

> Specifying  this  flag  automatically implies CLONE_FS as well.

dalehamel commented on src/utils.cpp:
```diff
@@ -521,13 +567,70 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
+}
+
+/*
+Determines if the target process is in the same mount namespace as bpftrace.
```
Good catch, I renamed the function (used to be `in_same_mountns`) and didn't change the comment.

dalehamel commented on src/utils.cpp:
```diff
@@ -618,15 +616,15 @@ static bool pid_in_different_mountns(int pid)
 
   sprintf(pid_path, "/proc/%d/ns/mnt", pid);
-  if(access( pid_path, F_OK ) != -1) {
+  if (access(pid_path, F_OK) != -1) {
```
Code was removed

dalehamel commented on tests/testprogs/mountns_wrapper.c:
```diff
@@ -0,0 +1,78 @@
+  if (mount(dname, private_mount, NULL, MS_BIND, NULL) != 0)
+    errExit("Failed to set up private bind mount");
+
+  if ((strlen(private_mount) + strlen(argv[1])) > (BUFFER_SIZE - 1))
```
Looking more closely, I realized I can just drop this check altogether and use `snprintf`, I also changed `BUFFER_SIZE` to `PATH_MAX`, as we are actually concatenating arbitrary paths here.

dalehamel commented:
thanks for the review comments @fbs @acj I believe I've addressed them all

fbs commented on src/utils.cpp:
```diff
@@ -521,13 +566,70 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
+
+  if (errno == EACCES || errno == ENOENT || bytes > BUFSIZE) {
+    std::cerr << "Error reading mount ns from " << pid_path;
+    std::cerr << " for pid " << pid << std::endl;
```
can you add the error here too? `strerror(errno)` iirc

dalehamel replied:
Done, I noticed this is failing on CI, I wonder if it will show a permission error on the next run...

dalehamel replied:
I figured out the issue, comment below.

dalehamel commented:
@fbs I found the reason why it was failing. When I removed the call to `access` to check if the file exists, using `readlink` will give `ENOENT`, because the symlink is (by design) a dangling symlink.

I decided to ditch the approach of using `readlink` at all, and looked to see how bcc does this, here is a great example:

https://github.com/iovisor/bcc/blob/0aca2c71c9b9e6b3d3335b6df288ad01a09c3ed6/src/cc/libbpf.c#L881-L932

I adapted this code, dropping the GOTO, but preserving the behavior. This just reads and compares the inode numers directly. CI now passes.

As a bonus, using `perror` should help with reporting errors.

dalehamel commented on src/utils.cpp:
```diff
@@ -521,13 +575,92 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
+  if (self_fd < 0)
+  {
+    perror("open(/proc/self/ns/mnt)");
+    close(self_fd);
```
I've gone with the goto suggestion, which I think handles this better.

dalehamel commented on src/utils.cpp:
```diff
@@ -521,13 +575,92 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
+
+If an invalid PID is specified or doesn't exist, it returns false.
+True is only returned if the namespace of the target process could be read and
+it doesn't match that of bpftrace.
```
I decided to add an exception, currently not catching it. I think that it will be more helpful than these errors.

Let me know what you think.

dalehamel replied:
I'll need to update the comment if you like the approach - reminder for myself.

fbs replied:
Looks ok to me, the same approach is used for WildcardException

fbs commented on src/utils.cpp:
```diff
@@ -521,13 +575,101 @@ std::vector<std::string> resolve_binary_path(const std::string& cmd)
+
+  if (fstat(target_fd, &target_stat))
+  {
+    errmsg = "fstat(target_fd)";
```
Knowing why it failed can be useful

```suggestion
    errmsg << "fstat(target_fd): " << strerror(errno);
```

(need to make errmsg a stringstream for this)

dalehamel replied:
Added thanks for the suggestion

fbs commented on src/utils.cpp:
```diff
@@ -508,10 +513,59 @@ std::string exec_system(const char* cmd)
+  std::string env_paths = "";
+  std::ostringstream pid_environ_path;
+
+  if (pid_in_different_mountns(pid))
```
```suggestion
  if (pid > 0 && pid_in_different_mountns(pid))
```

dalehamel commented:
@fbs thanks for another round of reviews, I believe I've addressed your comments.

If it looks good I'd be happy to squash/rebase.

dalehamel commented:
squashed+rebased
